src/snowflake/snowpark/modin/plugin/compiler/snowflake_query_compiler.py

```diff
@@ -6061,79 +5958,77 @@ def apply(
             Keyword arguments to pass to `func`.
         """
 
-        # axis=0 will always be processed with stored procedure fallback.
+        # axis=0 is not supported, raise error.
+        if axis == 0:
+            ErrorMessage.not_implemented(
+                "Snowpark pandas apply API doesn't yet support axis == 0"
+            )
         # Only callables are supported for axis=1 mode for now.
-        require_fallback = (
-            (not callable(func) and not isinstance(func, UserDefinedFunction))
-            or axis == 0
-            or result_type is not None
-            or check_snowpark_pandas_object_in_arg(args)
-            or check_snowpark_pandas_object_in_arg(kwargs)
-        )
-        if require_fallback:
-            return DataFrameDefault.register(native_pd.DataFrame.apply)(
-                self,
-                func=func,
-                axis=axis,
-                raw=raw,
-                result_type=result_type,
-                args=args,
-                **kwargs,
+        if not callable(func) and not isinstance(func, UserDefinedFunction):
```

Comment:
Really nice that now we have a separate error message for each reason why a NotImplementedError is raised.